naming-conventions.md

```diff
@@ -0,0 +1,188 @@
+```
+
+`m` is almost always `Monad` or something monad-transformer related
+```haskell
+when :: Monad m => Bool -> m () -> m ()
+liftIO :: MonadIO m => IO a -> m a
+```
+
+`s` is for state and `e` is for environment
+```haskell
+put :: MonadState s m => s -> m ()
+ask :: MonadReader e m => m e
+```
+
+These are conventions only, but picking something unusual for a common typeclass might surprise other readers.
+
+### Concrete Types
+
+Avoid using single-letter names or abbreviations for values with a concrete type, especially if that type is domain-specific. To crib from the [Swift API Design Guidelines](https://swift.org/documentation/api-design-guidelines/#fundamentals), "Clarity is more important than brevity."
+
+For example, don't do this:
+```haskell 
+fetchCourseMemberships :: MonadIO m => Entity Teacher -> SqlReadT m [Entity CourseMembership]
+fetchCourseMemberships (Entity tId teacher) =
+  select $ from $ \(ts `InnerJoin` cs `InnerJoin` cms) -> do
+    on $ cs ^. CourseId ==. cms  ^. CourseMembershipCourseId
+    on $ cs ^. CourseTeacherId ==. ts ^. TeacherId
+    where_ $
+      case teacherSchoolId teacher of
+        Just sId ->
+          ts ^. TeacherSchoolId ==. val (Just sId)
+        Nothing ->
+          ts ^. TeacherId ==. val tId
+    pure cms
+```
+
+Do this instead:
+```haskell
+fetchCourseMemberships :: MonadIO m => Entity Teacher -> SqlReadT m [Entity CourseMembership]
+fetchCourseMemberships (Entity teacherId teacher) =
+  select $ from $ \(teachers `InnerJoin` courses `InnerJoin` memberships) -> do
+    on $ courses ^. CourseId ==. memberships  ^. CourseMembershipCourseId
+    on $ courses ^. CourseTeacherId ==. teachers ^. TeacherId
+    where_ $
+      case teacherSchoolId teacher of
+        Just schoolId ->
+          teachers ^. TeacherSchoolId ==. val (Just schoolId)
+        Nothing ->
+          teachers ^. TeacherId ==. val teacherId
+    pure memberships
+```
+
+We tolerate a few abbreviations as components of identifiers (e.g. the `num` portion of `numQuestionsAnswered`), but there doesn't seem to be any broad consensus about this.
+
+## JavaScript
+
+JavaScript uses `camelCase` for identifiers and `TitleCase` for classes and React components. For flow types, we generally append a `T` to the end of the type, e.g. `SkillT`. Names that are being ignored should be prefixed with an underscore (e.g. `_teacher`, `_standard`). It's a common mistake (especially for Haskellers) to use a lone `_` and accidentally overwrite underscore in that scope.
+
+## {DRAFT} Interoperation Between Languages
+
+JSON keys should always be `camelCase`. String enums that cross language barriers should be `snake_case`, since most enums that cross language barriers are designed for `postgres` first. Exceptions can be made for short initialisms which may be `ALLCAPS` (e.g. `TEKS: MathSystemT`). Short initialisms that are part of other identifiers should use `snake_case` (e.g. `rti_coordinator: TeacherRoleT`).
```

Comment:
Not sure I agree that string enums that corss boundaries should be `snake_case`. What makes this necessary? We have separate encoding/decoding functions for postgres/json. We should be able to support each environments style.

Reply:
We have separate encoding/decoding functions for postgres/json, but we also have JSON that's stored in the DB with embedded enums (some of which are already `snake_case`).

Reply:
I guess this is the difference between this doc being inspirational or consistent with existing style?

Reply:
Yeah, I don't think there is a way to be consistent with our existing style :stuck_out_tongue: